feat: Adding the Terraform module (#46)
canonical/sdcore-nms-k8s-operator: terraform/README.md

# SD-Core NMS K8s Terraform Module

This SD-Core NMS K8s Terraform module aims to deploy the [sdcore-nms-k8s charm](https://charmhub.io/sdcore-nms-k8s) via Terraform.

## Getting Started

### Prerequisites

The following software and tools needs to be installed and should be running in the local environment. Please [set up your environment](https://discourse.charmhub.io/t/set-up-your-development-environment-with-microk8s-for-juju-terraform-provider/13109) before deployment.

- `microk8s`
- `juju 3.x`
- `terrafom`

### Module structure

- **main.tf** - Defines the Juju application to be deployed.
- **variables.tf** - Allows customization of the deployment. Except for exposing the deployment options (Juju model name, channel or application name) also models the charm configuration.
- **output.tf** - Responsible for integrating the module with other Terraform modules, primarily by defining potential integration endpoints (charm integrations), but also by exposing the application name.
- **terraform.tf** - Defines the Terraform provider.

> ## Using Sdcore-nms-k8s base module in higher level modules

Reviewer: ```suggestion
## Using sdcore-nms-k8s base module in higher level modules
```

Author: Done

If you want to use `sdcore-nms-k8s` base module as part of your Terraform module, import it like shown below.

```text
module "sdcore-nms-k8s" {
  source                 = "git::https://github.com/canonical/sdcore-nms-k8s-operator//terraform"
  model_name             = "juju_model_name"  
  # Optional Configurations
  # channel                        = "put the Charm channel here" 
}
```

Create the integrations, for instance:

```text
resource "juju_integration" "nms-sdcore-management" {
  model = var.model_name

  application {
    name     = module.nms.app_name

>     endpoint = "sdcore-management"

Reviewer: ```suggestion
    endpoint = module.nms.sdcore_management_endpoint
```

> # SD-Core NMS K8s Terraform Module

Author: I do not agree here. This module provides the guidance to deploy the NMS app and how to set up necessary relations.  Deploying the dependency apps should not be this module's responsibility.  The users are responsible to provide proper applications to relate.

> # SD-Core NMS K8s Terraform Module

Reviewer: In that case, should we mention the dependencies in the `Prerequisites` section?
Also if it's just for the NMS, I think we don't need to enable `multus` nor `metallb` addons.

> # SD-Core NMS K8s Terraform Module

Author: Multus and metallb plugins are removed from the requirements.

> # SD-Core NMS K8s Terraform Module

Author: Adding the dependent applications in the Prerequisites.

This SD-Core NMS K8s Terraform module aims to deploy the [sdcore-nms-k8s charm](https://charmhub.io/sdcore-nms-k8s) via Terraform.

## Getting Started

### Prerequisites

The following software and tools needs to be installed and should be running in the local environment.

- `microk8s`
- `juju 3.x`
- `terrafom`

### Deploy the sdcore-nms-k8s charm using Terraform

Make sure that `storage`, `multus` and `metallb` plugins are enabled for Microk8s:

```console
sudo microk8s enable hostpath-storage multus
sudo microk8s enable metallb:10.0.0.2-10.0.0.4
```

Add a Juju model:

```console

> juju add model <model-name>

Reviewer: ```suggestion
juju add-model <model-name>
```

Author: Done

Author: Good catch :)

```

Initialise the provider:

```console
terraform init
```

> Customize the configuration inputs under `terraform.tfvars` file according to requirement.
> 
> Replace the values in the `terraform.tfvars` file:

Reviewer: ```suggestion
Fill in the mandatory config options in the `terraform.tfvars` file:
```
I feel like having both these lines is a bit redundant. WDYT?

Author: I think we can replace.

```yaml
# Mandatory Config Options
model_name               = "put your model-name here"

> traefik_application_name = "put your nrf app name here"

Reviewer: ```suggestion
traefik_application_name = "put your Traefik app name here"
```

Author: Done

> webui_application_name   = "put your nrf app name here"

Reviewer: ```suggestion
webui_application_name   = "put your webui app name here"
```

Author: Done

```

> Run Terraform Plan by providing var-file:

Reviewer: ```suggestion
Create Terraform plan:
```

Author: Done

```console
terraform plan -var-file="terraform.tfvars" 
```

> Deploy the resources, skip the approval:

Reviewer: ```suggestion
Deploy the resources:
```

Author: I think it's not bad to say `Deploy the resources by skipping the approval`

Reviewer: I was mostly concerned about the `by` (which I see is gone now). `By` suggests that it's skipping the approval that deploys the resources, which is not true. Issuing the `terraform apply` command deploys the resources. Skipping the approval is just an optional argument. That's why I suggested to remove this part completely.

Author: `skip approval` is removed.

```console
terraform apply -auto-approve 
```

### Check the Output

Run `juju switch <juju model>` to switch to the target Juju model and observe the status of the applications.

```console
juju status --relations
```

### Clean up

> Remove the applications:

Reviewer: ```suggestion
Destroy the deployment:
```

Author: Done
